### ISSUE TITLE
Is this line used to calculate softmax probabilities for all neighbor nodes with same type?

### ISSUE THREAD

**Issue by chunyuma** [ZacharyChenpk/HetSANN#1]
Hello @ZacharyChenpk, I like your pytorch-version code to implement the HetSANN. But I'm a little bit confused about one line code [here](https://github.com/ZacharyChenpk/HetSANN/blob/e4b5836a7fa5f0bd3e47cd331c4d9da9944c0a25/model.py#L68). It seems like this line is used to calculate something like the softmax probabilities for all neighbor nodes with same type. But I'm not sure because the sum of `dim=1 ` of attn_o is not equal to 1 by using `attn_o - attn_o.max()`. Could you please help me explain why it has to calculate the attention by this way? Thank you so much!

**Comment by ZacharyChenpk** [comment]
In my mind, attn_o means exp(o_e^{l+1,m}) in Eq.4. The calculation of attention is cross all types of incoming relations (all from_type), and for convenience in coding, I implement this part with attention_dst_sum in line 70-72. The code 'ret[i]/(attention_dst_sum[i].unsqueeze(2)+1e-9)' in line 75 ensure the sum of attention scores is equal to 1. If you plug Eq.4 into Eq.5 and calculate the numerator (ret) and denominator (attention_dst_sum) respectively, it will be easier to understand.
And the ' - attn_o.max()' is just for avoiding the exp explosion, unrelated to the paper algorithm.
I have written this code months ago and don't remember the details clearly. If anything in my explanation is wrong, figure it out plz.

**Comment by chunyuma** [comment]
Ha, I see! Thanks so much for your detailed explanation @ZacharyChenpk! Your code is excellent!

One suggestion is to add `torch.nn.LeakyReLU` to the result of this [line](https://github.com/ZacharyChenpk/HetSANN/blob/e4b5836a7fa5f0bd3e47cd331c4d9da9944c0a25/model.py#L65) based on the Eq. 2 which has a LeakyReLU activation function.
